mathesar/imports/csv.py

```diff
@@ -1,15 +1,141 @@
-import csv
 from io import TextIOWrapper
 
+import clevercsv as csv
+
 from mathesar.database.base import create_mathesar_engine
 from mathesar.database.utils import get_database_key
 from mathesar.models import Table, Schema
+from mathesar.errors import InvalidTableError
 from db import tables, records
 
+ALLOWED_DELIMITERS = ",\t:| "
+SAMPLE_SIZE = 20000
+CHECK_ROWS = 10
+
+
+def parse_row(row, delimiter, escapechar, quotechar):
+    """
+    Parses a row of a *sv file given a set of dialetic parameters
+
+    Tries to mimic how PostgreSQL parses CSV files:
+        - Escape characters only escape quotes
+        - Escape characters only work inside quotes
+        - If no escape character is specified, it is the quote character
+
+    Args:
+        row: str, the row of text to parse
+        delimiter: str, a single character that divides the fields of the row
+        escapechar: str, a single character that escapes quotes
+        quotechar: str, a single character that is the type of quote uses
+
+    Returns:
+        splits: None if the row was invalid (due to an unmatched quote), otherwise a
+        list of lists, where the inner list contains every character of the field.
+
+    """
+    in_quote = False
+    escaped = False
+    row_len = len(row)
+    splits = []
+    current_split = []
+
+    i = 0
+    while i < row_len:
+        current_char = row[i]
+        if current_char == quotechar and not escaped:
+            if (in_quote and not escapechar
+                    and i != row_len - 1 and row[i + 1] == quotechar):
+                # Handle double quote escapes when there is no escape char
+                # Escapes are only respected within quotes
+                current_split.append(current_char)
+                i += 1
+            else:
+                in_quote = not in_quote
+        elif current_char == escapechar and in_quote and not escaped:
+            # Escapes are only respected within quotes
+            escaped = True
+        elif current_char == delimiter and not in_quote:
+            splits.append(current_split)
+            current_split = []
+        else:
+            current_split.append(current_char)
+            escaped = False
+        i += 1
+
+    if in_quote:
+        return None
+
+    splits.append(current_split)
+    return splits
+
+
+def check_dialect(file, dialect):
```

Comment:
Did you consider using the `stream_table` function from `clevercsv`?  

https://clevercsv.readthedocs.io/en/latest/source/clevercsv.html#clevercsv.wrappers.stream_table

It would be easier to maintain than the custom parser, and I'd much prefer that unless there's a good reason to roll our own here.